Tarefa_001_Marcos_Garcia

Definição do problema (em Markdown no topo do seu notebook)\
Escreva um pequeno “contrato do projeto” respondendo:

Dataset escolhido\
Penguins\
\
Qual é o target (y) que você quer prever\
Meu target (y) é o Species (Espécies)\
\
O problema é classificação\
Tentar classificar os penguins nas expécies existentes neste DataSet\
\
Quais colunas serão usadas como features (X)\
Onde ele vive (island)\
O tamanho do bico (bill_length_mm, bill_depth_mm)\
O tamanho da nadadeira (flipper_length_mm)\
O peso (body_mass_g)\
O sexo (sex)\
\
Quais colunas você vai remover e por quê (ex.: colunas inúteis, texto livre, identificadores)\
Não removerei nenhua coluna

In [15]:
# 1) Carregar o dataset Penguins
df_t = sns.load_dataset("penguins")

display(df_t.head())
print("Shape:", df_t.shape)

print("\nTipos das colunas:")
print(df_t.dtypes)

print("\nFaltantes por coluna (top 10):")
display(df_t.isna().sum().sort_values(ascending=False).head(10))

print("\nDistribuição do target (species):")
display(df_t["species"].value_counts())
display(df_t["species"].value_counts(normalize=True).rename("Espécies"))

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


Shape: (344, 7)

Tipos das colunas:
species               object
island                object
bill_length_mm       float64
bill_depth_mm        float64
flipper_length_mm    float64
body_mass_g          float64
sex                   object
dtype: object

Faltantes por coluna (top 10):


,0
sex,11
bill_depth_mm,2
bill_length_mm,2
flipper_length_mm,2
body_mass_g,2
island,0
species,0



Distribuição do target (species):


,count
species,
Adelie,152
Gentoo,124
Chinstrap,68


,Espécies
species,
Adelie,0.441860
Gentoo,0.360465
Chinstrap,0.197674


In [30]:
#Informações da estrutura do DataFrame "penguins"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [17]:
# 2) Separar X e y
y = df_t["species"]
X = df_t.drop(columns=["species"])

print("X shape:", X.shape)
print("y shape:", y.shape)
display(X.head())

X shape: (344, 6)
y shape: (344,)


,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Torgersen,36.7,19.3,193.0,3450.0,Female


In [18]:
# 3) Identificar colunas por tipo
num_cols = X.select_dtypes(include=["int64", "float64"]).columns
cat_cols = X.select_dtypes(include=["object", "category", "bool"]).columns

print("Numéricas:", list(num_cols))
print("Categóricas:", list(cat_cols))

Numéricas: ['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']
Categóricas: ['island', 'sex']


RandomForestClassifier: É um dos modelos mais populares e robustos. Ele constrói várias árvores de decisão e combina suas previsões, o que o torna menos propenso a overfitting e geralmente mais preciso que uma única árvore ou a Regressão Logística.

GradientBoostingClassifier: Outro modelo de conjunto muito poderoso que constrói árvores de decisão sequencialmente, onde cada nova árvore tenta corrigir os erros das árvores anteriores. Ele tende a ter um desempenho excelente em muitos conjuntos de dados.

SVC (Support Vector Classifier): Máquinas de Vetores de Suporte podem ser muito eficazes, especialmente em dados com boa separabilidade. Com o uso de kernels (como o RBF), podem modelar relações não-lineares complexas, o que a Regressão Logística linear não consegue.

KNeighborsClassifier: Embora seja um modelo mais simples, baseado na distância entre os pontos de dados, pode ser uma melhoria significativa se os dados tiverem fronteiras de decisão não-lineares que a Regressão Logística não consegue capturar. É fácil de entender e implementar.

Modelos (baseline + melhoria) (obrigatório)
Treine e compare dois modelos, mudando somente o classificador (para a comparação ser justa):

Baseline (obrigatório): um modelo simples e interpretável\
Melhoria simples (obrigatório): um modelo mais potente\
Explique em 2–3 linhas por que você considerou um “baseline” e o outro “melhoria”.\
A BaseLine escolhida foi a LogisticRegression pois foi a que tinha no modelo usado de exemplo feito pelo professor, em minhas outras tentativas de 3 melhorias, a única que igualou-se a Accuracy da BaseLine foi a terceira na qual usei SVM para o treinamento.

In [22]:
# 4) Dividir em treino e teste (reprodutível e com estratificação do target)
RANDOM_STATE = 50
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=RANDOM_STATE,
    stratify=y
)

print("Treino:", X_train.shape, "Teste:", X_test.shape)
print("Proporção do target (treino):")
display(y_train.value_counts(normalize=True).rename("proporção"))
print("Proporção do target (teste):")
display(y_test.value_counts(normalize=True).rename("proporção"))

Treino: (275, 6) Teste: (69, 6)
Proporção do target (treino):


,proporção
species,
Adelie,0.443636
Gentoo,0.360000
Chinstrap,0.196364


Proporção do target (teste):


,proporção
species,
Adelie,0.434783
Gentoo,0.362319
Chinstrap,0.202899


In [32]:
def avaliar_classificacao(nome, y_true, y_pred):
    acc = accuracy_score(y_true, y_pred)
    cm = confusion_matrix(y_true, y_pred)

    print(f"\n=== {nome} ===")
    print(f"Accuracy: {acc:.4f}")
    print("\nClassification report:")
    print(classification_report(y_true, y_pred))
    print("Matriz de confusão:")
    print(cm)

    # Interpretação para classificação multiclasse
    print("\nInterpretação da Matriz de Confusão (Multiclasse):")
    print("- A diagonal principal mostra o número de predições corretas para cada classe.")
    print("- Os valores fora da diagonal principal mostram os erros do modelo, ou seja, onde o modelo previu uma classe diferente da verdadeira.")

    return {
        "acc": acc,
        "cm": cm
    }

In [31]:
#BaseLine
model_baseline_lr = Pipeline(steps=[
    ("prep", preprocess),
    ("clf", LogisticRegression(max_iter=1000, random_state=RANDOM_STATE))
])

print("Treinando baseline (LogisticRegression)...")
model_baseline_lr.fit(X_train, y_train)

pred_baseline_lr = model_baseline_lr.predict(X_test)

res_lr = avaliar_classificacao("Baseline — LogisticRegression", y_test, pred_baseline_lr)
acc_baseline_lr = res_lr["acc"]


Treinando baseline (LogisticRegression)...

=== Baseline — LogisticRegression ===
Accuracy: 0.9855

Classification report:
              precision    recall  f1-score   support

      Adelie       0.97      1.00      0.98        30
   Chinstrap       1.00      0.93      0.96        14
      Gentoo       1.00      1.00      1.00        25

    accuracy                           0.99        69
   macro avg       0.99      0.98      0.98        69
weighted avg       0.99      0.99      0.99        69

Matriz de confusão:
[[30  0  0]
 [ 1 13  0]
 [ 0  0 25]]

Interpretação da Matriz de Confusão (Multiclasse):
- A diagonal principal mostra o número de predições corretas para cada classe.
- Os valores fora da diagonal principal mostram os erros do modelo, ou seja, onde o modelo previu uma classe diferente da verdadeira.


In [36]:
#Melhoria1
from sklearn.ensemble import GradientBoostingClassifier

model_baseline_lr = Pipeline(steps=[
    ("prep", preprocess),
    ("clf", GradientBoostingClassifier(random_state=RANDOM_STATE))
])

print("Treinando modelo (GradientBoostingClassifier) como melhoria...")
model_baseline_lr.fit(X_train, y_train)

pred_baseline_lr = model_baseline_lr.predict(X_test)

res_lr = avaliar_classificacao("Melhoria — GradientBoostingClassifier", y_test, pred_baseline_lr)
acc_baseline_lr = res_lr["acc"]

Treinando modelo (GradientBoostingClassifier) como melhoria...

=== Melhoria — GradientBoostingClassifier ===
Accuracy: 0.9565

Classification report:
              precision    recall  f1-score   support

      Adelie       0.97      0.93      0.95        30
   Chinstrap       0.93      0.93      0.93        14
      Gentoo       0.96      1.00      0.98        25

    accuracy                           0.96        69
   macro avg       0.95      0.95      0.95        69
weighted avg       0.96      0.96      0.96        69

Matriz de confusão:
[[28  1  1]
 [ 1 13  0]
 [ 0  0 25]]

Interpretação da Matriz de Confusão (Multiclasse):
- A diagonal principal mostra o número de predições corretas para cada classe.
- Os valores fora da diagonal principal mostram os erros do modelo, ou seja, onde o modelo previu uma classe diferente da verdadeira.


In [37]:
#Melhoria2
from sklearn.ensemble import RandomForestClassifier

model_baseline_lr = Pipeline(steps=[
    ("prep", preprocess),
    ("clf", RandomForestClassifier(random_state=RANDOM_STATE))
])

print("Treinando modelo (RandomForestClassifier) como melhoria...")
model_baseline_lr.fit(X_train, y_train)

pred_baseline_lr = model_baseline_lr.predict(X_test)

res_lr = avaliar_classificacao("Melhoria — RandomForestClassifier", y_test, pred_baseline_lr)
acc_baseline_lr = res_lr["acc"]

Treinando modelo (RandomForestClassifier) como melhoria...

=== Melhoria — RandomForestClassifier ===
Accuracy: 0.9710

Classification report:
              precision    recall  f1-score   support

      Adelie       0.97      0.97      0.97        30
   Chinstrap       0.93      0.93      0.93        14
      Gentoo       1.00      1.00      1.00        25

    accuracy                           0.97        69
   macro avg       0.97      0.97      0.97        69
weighted avg       0.97      0.97      0.97        69

Matriz de confusão:
[[29  1  0]
 [ 1 13  0]
 [ 0  0 25]]

Interpretação da Matriz de Confusão (Multiclasse):
- A diagonal principal mostra o número de predições corretas para cada classe.
- Os valores fora da diagonal principal mostram os erros do modelo, ou seja, onde o modelo previu uma classe diferente da verdadeira.


In [40]:
#Melhoria3
from sklearn.svm import SVC

model_baseline_lr = Pipeline(steps=[
    ("prep", preprocess),
    ("clf", SVC(random_state=RANDOM_STATE))
])

print("Treinando modelo (SVC) como melhoria...")
model_baseline_lr.fit(X_train, y_train)

pred_baseline_lr = model_baseline_lr.predict(X_test)

res_lr = avaliar_classificacao("Melhoria — SVC", y_test, pred_baseline_lr)
acc_baseline_lr = res_lr["acc"]

Treinando modelo (SVC) como melhoria...

=== Melhoria — SVC ===
Accuracy: 0.9855

Classification report:
              precision    recall  f1-score   support

      Adelie       0.97      1.00      0.98        30
   Chinstrap       1.00      0.93      0.96        14
      Gentoo       1.00      1.00      1.00        25

    accuracy                           0.99        69
   macro avg       0.99      0.98      0.98        69
weighted avg       0.99      0.99      0.99        69

Matriz de confusão:
[[30  0  0]
 [ 1 13  0]
 [ 0  0 25]]

Interpretação da Matriz de Confusão (Multiclasse):
- A diagonal principal mostra o número de predições corretas para cada classe.
- Os valores fora da diagonal principal mostram os erros do modelo, ou seja, onde o modelo previu uma classe diferente da verdadeira.


Comparativo de Accuracy:\
Baseline:0.9855\
Melhoria1:0.9565\
Melhoria2:0.9710\
Melhoria3:0.9855

Aqui está a parte mais importante da atividade.

Escreva uma interpretação em linguagem humana:

O que significa um FP no seu problema?\
Significa que o modelo previu que um pinguin da espécie A foi classificado como um pinguin de espécie C quando na verdade ele é da espécie A.\
\
O que significa um FN no seu problema?\
Significa que um penguin não foi classificado como espécie C e foi erroneamente classificado como pertencente a espécie A\
\
Qual erro é mais “grave” no contexto do dataset escolhido?\
O modelo baseline errou mais em qual classe? E o modelo melhorado?\

1- Pense nas implicações do mundo real. Existe alguma consequência diferente se um pinguim Gentoo é confundido com um Adelie, versus um Adelie ser confundido com um Chinstrap? Por exemplo, se uma espécie fosse rara ou estivesse em perigo, um Falso Negativo para essa espécie (não identificá-la quando ela está lá) poderia ser mais grave para os esforços de conservação.
Sem um contexto específico sobre as implicações de cada erro para o estudo ou conservação dos pinguins, você pode argumentar que, do ponto de vista puramente estatístico (e neste dataset simples), todos os erros de classificação são igualmente indesejáveis, pois o objetivo é identificar corretamente a espécie.

2- Conclusão para Melhoria 3 (SVC): O modelo SVC também cometeu o mesmo único erro, classificando um pinguim Chinstrap como Adelie. Isso significa que, neste conjunto de teste, o modelo SVC não demonstrou uma melhoria nos padrões de erro em relação ao baseline; ambos atingiram a mesma acurácia e cometeram o mesmo engano.

3- Porem, entre as supostas Melhorias 1 e 2 quem cometeu mais erros foi a Melhoria 1. Sendo que, "Melhoria 1" (GradientBoostingClassifier) cometeu 3 erros, enquanto a "Melhoria 2" (RandomForestClassifier) cometeu 2 erros.